// web-ifc-viewer/dist/components/display/edges/ifcEdges.js
import { LineSegments, LineBasicMaterial, EdgesGeometry } from 'three';
import { IfcComponent } from '../../../base-types';
export class IfcEdges extends IfcComponent {
    constructor(context) {
        super(context);
        this.lineMaterial = new LineBasicMaterial({
            color: 0x888888
        });
        this.toggleEdges = (modelIDs, active) => {
            this.context.items.ifcModels.forEach((ifcModel) => {
                if (!modelIDs.includes(ifcModel.modelID))
                    return;
                if (!active && ifcModel.userData.wireframe) {
                    ifcModel.remove(ifcModel.userData.wireframe);
                    return;
                }
                if (!ifcModel.userData.wireframe)
                    ifcModel.userData.wireframe = this.getEdges(ifcModel);
                ifcModel.add(ifcModel.userData.wireframe);
            });
        };
        this.vectorize = (item) => {
            const geometry = new EdgesGeometry(item.geometry);
            return new LineSegments(geometry, this.lineMaterial);
        };
        this.context = context;
    }
    set edgesMaterial(newMaterial) {
        this.lineMaterial = newMaterial;
    }
}
